src/main.cpp

```diff
@@ -1065,16 +1065,29 @@ bool CheckZerocoinSpend(const CTransaction& tx, bool fVerifySignature, CValidati
 
     bool fValidated = false;
     set<CBigNum> serials;
-    list<CoinSpend> vSpends;
     CAmount nTotalRedeemed = 0;
     for (const CTxIn& txin : tx.vin) {
 
         //only check txin that is a zcspend
-        if (!txin.scriptSig.IsZerocoinSpend())
+        bool isPublicSpend = txin.scriptSig.IsZerocoinPublicSpend();
+        if (!txin.scriptSig.IsZerocoinSpend() && !isPublicSpend)
             continue;
 
-        CoinSpend newSpend = TxInToZerocoinSpend(txin);
-        vSpends.push_back(newSpend);
+        CoinSpend newSpend;
+        CTxOut prevOut;
+        if (isPublicSpend) {
+            if(!GetOutput(txin.prevout.hash, txin.prevout.n, state, prevOut)){
```

Comment:
Can't we avoid code duplication here, re-using a slightly modified `ParseZerocoinPublicSpend` ?

Reply:
Can do some minor changes but still, we should have a CoinSpend cache like what we do with the utxo view in the future.